### ISSUE TITLE
hyphy relax error

### ISSUE THREAD

**Issue by bernt-matthias** [veg/hyphy#1265]
When executing: 

```
export OMP_NUM_THREADS="1" && hyphy relax --alignment relax-in1.fa --tree relax-in1.nhx --models 'All' --code 'Universal' --test TEST
```

using test data from https://github.com/galaxyproject/tools-iuc/tree/master/tools/hyphy/test-data I get:

```
Error:Internal error 
Constrained optimization failed, since a starting point within the domain specified for the variables couldn't be found.
Set it by hand, or check your constraints for compatibility.
Failed constraint:relax.ge.omega2:=0.02003649738120953/1e-08 must be in [1e-08,1]. Current value = 2003649.738120953.
Internal error 
Constrained optimization failed, since a starting point within the domain specified for the variables couldn't be found.
Set it by hand, or check your constraints for compatibility.
Failed constraint:relax.ge.omega2:=0.02003649738120953/1e-08 must be in [1e-08,1]. Current value = 2003649.738120953.%     
```

Any idea what could be wrong?

Sometimes I get an `Segmentation fault`.

Also run times vary quite a lot.

**Comment by spond** [comment]
Dear @bernt-matthias,

This is not an expected behavior, clearly. I was not able to replicate the issue within our build setups; I get the expected output when I try the same run (see below).

Can you:

1. Check the version of HyPhy (`hyphy --version`).
2. Tell me what system (OS, compiler you are testing it on)?
3. Run the command with ENV="VERBOSITY_LEVEL=10" added to the command line and report at what stage the program fails (you'll have a lot of stuff dumped to the console).

Best,
Sergei

Expected output:

```
./hyphy CPU=1 relax --alignment ~/Desktop/relax-in1.fa --tree ~/Desktop/relax-in1.nhx --test TEST                           


>code –> Universal

>Loaded a multiple sequence alignment with **10** sequences, **187** codons, and **1** partitions from `/Users/sergei/Desktop/relax-in1.fa`

>test –> TEST
The number of points in the initial distributional guess for likelihood fitting (permissible range = [1,10000], default value = 250, integer): 
>grid-size –> 250
The number of initial random guesses to 'seed' rate values optimization (permissible range = [1,25], default value = 1, integer): 
>starting-points –> 1


### Branch sets for RELAX analysis
* Selected 13 branches as the _Reference_ set: `Pig, Node3, Node2, RhMonkey, Baboon, Node9, Human, Chimp, Node12, Node8, Node1, Rat, Mouse`
* Selected 3 branches as the _Test_ set: `Cow, Horse, Cat`
The number omega rate classes to include in the model (permissible range = [2,10], default value = 3, integer): 
>rates –> 3

>models –> All

>srv –> No


### Obtaining branch lengths and nucleotide substitution biases under the nucleotide GTR model

>kill-zero-lengths –> Yes
* Log(L) = -3532.32, AIC-c =  7112.86 (24 estimated parameters)

### Obtaining the global omega estimate based on relative GTR branch lengths and nucleotide substitution biases
* Log(L) = -3465.92, AIC-c =  6996.99 (32 estimated parameters)
* non-synonymous/synonymous rate ratio for *Reference* =   0.9374
* non-synonymous/synonymous rate ratio for *Test* =   1.1194

### Improving branch lengths, nucleotide substitution biases, and global dN/dS ratios under a full codon model
* Log(L) = -3461.22, AIC-c =  6987.58 (32 estimated parameters)
* non-synonymous/synonymous rate ratio for *Reference* =   0.7691
* non-synonymous/synonymous rate ratio for *Test* =   1.6090

### Fitting the general descriptive (separate k per branch) model

### * Log(L) = -3420.63, AIC-c =  6946.17 (51 estimated parameters)
* The following baseline rate distribution for branch-site combinations was inferred

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.000     |   54.622    |                                   |
|        Negative selection         |     0.000     |    0.040    |       Collapsed rate class        |
|      Diversifying selection       |     2.206     |   45.338    |                                   |

* Branch-level relaxation or intensification parameter distribution has mean  1.78, median  1.48, and 95% of the weight in  0.15 -  5.57

### Fitting the alternative model to test K != 1
* Log(L) = -3429.23, AIC-c =  6931.92 (36 estimated parameters)
* Relaxation/intensification parameter (K) =     2.44
* The following rate distribution was inferred for **test** branches

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.000     |   50.188    |                                   |
|        Negative selection         |     0.000     |    3.285    |       Collapsed rate class        |
|      Diversifying selection       |     6.322     |   46.527    |                                   |

* The following rate distribution was inferred for **reference** branches

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.000     |   50.188    |                                   |
|        Negative selection         |     0.000     |    3.285    |       Collapsed rate class        |
|      Diversifying selection       |     2.129     |   46.527    |                                   |


### Fitting the null (K := 1) model
* Log(L) = -3437.33, AIC-c =  6946.04 (35 estimated parameters)
* The following rate distribution for test/reference branches was inferred

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.119     |   56.525    |                                   |
|        Negative selection         |     0.259     |    6.018    |                                   |
|      Diversifying selection       |     3.699     |   37.457    |                                   |

----
## Test for relaxation (or intensification) of selection [RELAX]
Likelihood ratio test **p =   0.0001**.
>Evidence for *intensification of selection* among **test** branches _relative_ to the **reference** branches at P<=0.05
----


### Fitting the partitioned descriptive model (completely separate rate distributions for branch sets)
* Log(L) = -3427.62, AIC-c =  6937.03 (40 estimated parameters)
* The following rate distribution was inferred for *test* branches 

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.000     |   51.008    |                                   |
|        Negative selection         |     0.000     |    0.467    |       Collapsed rate class        |
|      Diversifying selection       |     6.088     |   48.525    |                                   |

* The following rate distribution was inferred for *reference* branches 

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.190     |   36.681    |                                   |
|        Negative selection         |     0.274     |   40.413    |                                   |
|      Diversifying selection       |     4.170     |   22.906    |                                   |
```

**Comment by bernt-matthias** [comment]
Thanks for the fast feedback.

I'm running the hyphy docker container that is built from the bioconda recipe: https://quay.io/repository/biocontainers/hyphy?tab=tags (2.5.24--h3db2f75_0) .. the host OS is the latest LTS Ubuntu.

The compilation is done as follows: https://github.com/bioconda/bioconda-recipes/blob/master/recipes/hyphy/build.sh 

If I get the segfault the output stops  with:

```
### Fitting the partitioned descriptive model (completely separate rate distributions for branch sets)
* Log(L) = -3427.62, AIC-c =  6937.03 (40 estimated parameters)
* The following rate distribution was inferred for *test* branches 

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.000     |    2.882    |                                   |
|        Negative selection         |     0.000     |   48.527    |       Collapsed rate class        |
|      Diversifying selection       |     6.034     |   48.591    |                                   |

* The following rate distribution was inferred for *reference* branches 

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.223     |    5.461    |                                   |
|        Negative selection         |     0.230     |   71.328    |       Collapsed rate class        |
|      Diversifying selection       |     4.133     |   23.211    |                                   |
```

I'm currently trying to get debug output. Is it `VERBOSITY_LEVEL=10 && ./hyphy relax ... ` - this seems not to change the output?

**Comment by spond** [comment]
Dear @bernt-matthias,

Hmm, the only difference in the condo script and the default build script is to not use AVX vector extensions. I'll check with that flag on to see if the error is reproducible.

For the output, you need to specify `ENV="VERBOSITY_LEVEL=10;"` as a command line argument to HyPhy (it's an internal environment variable to HyPhy), so like 

```
./hyphy relax ENV="VERBOSITY_LEVEL=10;" ... 
```

Clearly this is an issue that should be caught by our tests, tagging @davebx who helps with Conda packages and testing. I'd like to get to the bottom of it and fix the issue. 

Best,
Sergei

**Comment by bernt-matthias** [comment]
Seems not to be reproducable. Sometimes I get no error, sometimes "just" a segfault and now I got once:

```
*** Error in `hyphy': corrupted double-linked list: 0x000055901f5a3110 ***
/tmp/tmpv5j_toma/job_working_directory/000/3/tool_script.sh: line 22:     8 Aborted                 (core dumped) hyphy relax --alignment ./relax_input.fa --tree ./relax_input.nhx --models 'All' --code 'Universal' --test TEST ENV="VERBOSITY_LEVEL=10;" > '/tmp/tmpv5j_toma/files/8/3/f/dataset_83fe1e4a-e6cd-4c60-a20e-ea5f4183390b.dat'
```

I also do not get any different output when using `ENV="VERBOSITY_LEVEL=10;"`

**Comment by spond** [comment]
Dear @bernt-matthias,

Thanks for letting me know. The segfaults are post-error conditions (sometimes HyPhy is not able to clean everything properly during an abnormal exit). I'll install the package from `Conda` and test it on CentOS and OS X.  

Do you encounter errors with other tests, or just this one?

Best,
Sergei

**Comment by bernt-matthias** [comment]
> Do you encounter errors with other tests, or just this one?

Only this one. And as far as I have seen the error(s) seem(s) to occur only when running in the container.

**Comment by spond** [comment]
Dear @bernt-matthias,

HyPhy 2.5.25 has been released here and on Conda; can you please check to see if the issue has been resolved?

Best,
Sergei

**Comment by bernt-matthias** [comment]
Unfortunately the problem persists.

**Comment by spond** [comment]
Dear @bernt-matthias,

OK, I will continue investigating! Thanks for reporting. 

Best,
Sergei

**Comment by spond** [comment]
Dear @bernt-matthias,

Hopefully fixed with 2.5.26.

Best,
Sergei

**Comment by bernt-matthias** [comment]
Looks good https://github.com/galaxyproject/tools-iuc/pull/3430

Thanks a lot
